### ISSUE TITLE
LVS violation on user_project_wrapper

### ISSUE THREAD

**Issue by boehmerst** [phthinh/soric_project#1]
I try to generate a user_project which includes several hard macros including SRAM. As your project is close to what I plan I used it to study the openlane flow and the caravel wrapper. I successfully hardened the various macros. However, the flow fails on the user_project_wrapper with 5 routing violations it cannot resolve and finally fails with LVS errors (shorts between wb_clock_i and vssd1 electrically shortened).

I wonder if this related to the OpenLane version that I am using. Can you provide me the version of the OpenLane, PDK and Caravel scripts you are using to get a DRC, LVS and timing clean top level gds.

Thanks you very much

Stephan

**Comment by phthinh** [comment]
Yeah, it sounds like the issue with the version of OpenLane and Caravel. In this project, I used the OpenLane at the version tag `tag: mpw-4` (commit: 4a1c79905681ea4c5a051d8c7e24038a7bc45af2), the caravel at the version tag `tag: mpw-3a` (commit: 4a1c79905681ea4c5a051d8c7e24038a7bc45af2), and the PDK at commit: 2a3c94373ffc0e53a287be94a569a2139b1cb496.

**Comment by boehmerst** [comment]
Thanks for your feedback. I pumped the OpenLane commit to the mpw-4 tag. Also I changed the tag inside the caravel/openlane/Makefile accordingly to use the mpw-4 tagged version of the docker container. Regarding the pdk I installed it using the OpenLane Makefile not the caravel Makefile. The commit seems to be still different from yours. The user_project_wrapper build fails with 90+ violations during the detailed routing step and finally with "[ERROR]: There are illegal overlaps (e.g., routes over obstructions) in your design."

As there are various ways to install the tools (caravel Makefiles vs. OpenLane Makefiles) and the caravel repos seemed to have passed significant refactoring resulting in the user_project_wrapper being incompatible.  May I ask you for  a step by step instruction on how to setup the flow based on your repo? I already tried 5 different designs that were taped out during the shuttle program but all of them fail at one step or another. Only simple designs with just one single macro pass the flow. So I suspect I am missing something fundamentally.

Thanks,
Stephan

**Comment by phthinh** [comment]
Hi Stephan,

Sorry for the mistake of the caravel commit. It should be: `c2943440e278814787f761585b99b9ea3c1f4121`. Could you check again your commits of Caravel, Openlane, and PDK are identical to that I shared?
If I remember correctly, I installed the PDK from Openlane Makefile in a native way that does not use Docker. Indeed, I also struggled with toolchains' incompatibility for weeks last time until I got to the suitable versions of Caravel, Openlane, and PDK that I already shared with you. Apart from that, I followed the instructions of the caravel_user_project repo.

**Comment by boehmerst** [comment]
Hi phthinh,

Thanks for your hint. I double checked the commits and setup the flow again. The flow still fails but now at the final magic DRC stage with roughly 5 million DRC errors. I read that magic DRC violations are very often fault alerts as long as the Triton router reports no violation. However, I am not sure about that. According to your signoff report you managed to pass the flow without violation, right?
There is still one inconsistency as the Makefile to run the flow is starting a docker image with MPW-3a tag while the OpenLane installation generates an MPW-4 tagged image. I will try to rerun the flow accordingly. 
My learning is that the OpenLane flow is still shaky and things can break easily. I think more complex designs are required within the continuous integration and releasing process. When I use the latest OpenLane version even the macro hardening fails.

**Comment by phthinh** [comment]
Hi Stephen,

Yeah, I was told that It's probably that magic DRC violations are very often having fault alerts. And I am not sure the local magic DRC check is properly updated. As you can see in the Openlane script for the user project wrapper, [openlane/user_project_wrapper/config.tcl](https://github.com/phthinh/soric_project/blob/master/openlane/user_project_wrapper/config.tcl), I disabled running Magic DRC (line 116) locally. But eventually, we have the Magic DRC at the Efabless portal (it has to be passed) and our GDS passed that check for the Precheck job.  I guess that you probably reproduced the GDS file for the Soric project. I look forwards to seeing your project in MPW-5.

**Comment by boehmerst** [comment]
Good point, I was indeed wondering why magic DRC check was performed actually. It seems like the RUN_MAGIC_DRC flag was removed. I rerun with MAGIC_DRC_USE_GDS flag to perform DRC checking on lef level. This results in a clean run without any magic DRC violations. As I presume the standard cells and SRAM macros to be DRC clean by definition I am optimistic that the DRC violations on GDS level are fault alerts.

Yes, I reproduced the GDS files for the Soric project for now to get a working setup as a starting point. Not sure if a get something done for the MPW-5 run already. I registered for the efabless webinar. I hope that they present something about the flow and tape out procedure especially since the user_project_wraper + caravel is currently broken and the tags for the tools got inconsistent. 
Anyway, looking forward to see you at MPW-X tape out parties! 😉 

Thanks for the discussion!
